flatnav/Index.h

```diff
@@ -57,28 +103,79 @@ template <typename dist_t, typename label_t> class Index {
   // this class.
   Index() = default;
 
-  ~Index() { delete[] _index_memory; }
+  ~Index() {
+    delete[] _index_memory;
+    delete _visited_set_pool;
+  }
 
-  void add(void *data, label_t &label, int ef_construction,
-           int num_initializations = 100) {
-    // initialization must happen before alloc due to a bug where
-    // initializeSearch chooses new_node_id as the initialization
-    // since new_node_id has distance 0 (but no links). The search is
-    // skipped because the "optimal" node seems to have been found.
-    node_id_t new_node_id;
-    node_id_t entry_node = initializeSearch(data, num_initializations);
+  /**
+   * @brief Add a new vector to the index.
+   * Initialization must happen before allocation due to a bug where
+   * initializeSearch chooses new_node_id as the initialization
+   * since new_node_id has distance 0 (but no links). The search is
+   * skipped because the "optimal" node seems to have been found.
+   *
+   * @param data                 The vector to add.
+   * @param label                The label (meta-data) of the vector.
+   * @param ef_construction      The search beam width. This is the `ef`
+   * parameter in the HNSW paper.
+   * @param num_initializations  The number of random initializations to use.
+   * This determines the number of nodes to search before adding the new node.
+   */
+  void addParallel(void *data, std::vector<label_t> &labels,
+                   int ef_construction, int num_initializations = 100) {
+    if (num_initializations <= 0) {
+      throw std::invalid_argument(
+          "num_initializations must be greater than 0.");
+    }
+
+    uint32_t total_num_nodes = labels.size();
+    uint32_t data_dimension = _distance->dimension();
+
+    // Don't spawn any threads if we are only using one.
+    if (_num_threads == 1) {
+      for (uint32_t row_id = 0; row_id < total_num_nodes; row_id++) {
+        void *vector = (float *)data + (row_id * data_dimension);
+        label_t label = labels[row_id];
+        concurrentAdd(vector, label, ef_construction, num_initializations);
+      }
+      return;
+    }
+
+    flatnav::executeInParallel(
+        /* start_index = */ 0, /* end_index = */ total_num_nodes,
+        /* num_threads = */ _num_threads, /* function = */
+        [&](uint32_t row_index) {
+          void *vector = (float *)data + (row_index * data_dimension);
+          label_t label = labels[row_index];
+          concurrentAdd(vector, label, ef_construction, num_initializations);
+        });
+  }
+
+  void concurrentAdd(void *data, label_t &label, int ef_construction,
```

Comment:
`addBatch` is kind of a misnomer since it implies that the `data` argument is a single batch, but I like it better than `addParallel`. I'll change this accordingly and add proper documentation.